stage0/src/dice_attestation.rs

```diff
@@ -0,0 +1,129 @@
+// limitations under the License.
+//
+
+extern crate alloc;
+
+use alloc::{string::String, vec::Vec};
+use coset::{cbor::value::Value, cwt, iana, CborSerializable, CoseError};
+use hkdf::Hkdf;
+use p384::ecdsa::{signature::Signer, Signature, SigningKey, VerifyingKey};
+use rand_core::OsRng;
+use sha2::Sha256;
+
+
+pub const ID_SALT: &str = "DICE_ID_SALT";
+pub const SIGNATURE_LENGTH: usize = 20;
+
+/// Attestation related functions
+///
+/// Returns Signed Stage1 key and measurments
+struct Stage0Signer {
+    signing_key: SigningKey,
+}
+
+// Signer implementation.
+impl Stage0Signer {
+    fn sign(&self, data: &[u8]) -> Vec<u8> {
+        let signed_stage1_ca_verifying_key: Signature = self.signing_key.sign(data);
+        Vec::from(signed_stage1_ca_verifying_key.to_bytes().as_slice())
+    }
+}
+
+fn derive_id(ikm: &[u8], info: Option<&[u8]>) -> [u8; SIGNATURE_LENGTH] {
+    let hkdf = Hkdf::<Sha256>::new(Some(ID_SALT.as_bytes()), ikm);
+    let mut okm: [u8; SIGNATURE_LENGTH] = [0u8; SIGNATURE_LENGTH];
+    hkdf.expand(info.unwrap_or(&[]), &mut okm)
+        .expect("20 is a valid length for Sha256 to output");
+
+    okm
+}
+
+fn generate_ecdsa_keys(info: &str) -> (SigningKey, [u8; SIGNATURE_LENGTH]) {
+    let key = SigningKey::random(&mut OsRng);
+    let public_key = VerifyingKey::from(&key);
+    let key_id = derive_id(
+        public_key.to_encoded_point(false).as_bytes(),
+        Some(info.as_bytes()),
+    );
+    (key, key_id)
+}
+
+fn generate_stage1_ca_cwt(
+    stage0_eca_key: SigningKey,
+    stage0_cert_id_hex: String,
+) -> Result<Vec<u8>, CoseError> {
+    // Generate Stage 1 keys and Signer.
+    let info_str = "ID";
+    let stage1_ca_key = generate_ecdsa_keys(info_str);
+    let stage1_ca_verifying_key = VerifyingKey::from(&stage1_ca_key.0);
+    let stage0_signer = Stage0Signer {
+        signing_key: stage0_eca_key,
+    };
+
+    // Generate claims
+    let claims = cwt::ClaimsSetBuilder::new()
+        .issuer(stage0_cert_id_hex)
+        .subject(hex::encode(stage1_ca_key.1))
+        .cwt_id(Vec::from(stage1_ca_key.1))
+        // Add additional private-use claim.
+        .private_claim(
+            -4670552,
+            Value::Bytes(Vec::from(
+                stage1_ca_verifying_key.to_encoded_point(false).as_bytes(),
+            )),
+        )
+        .build();
+
+    let aad = b"";
+
+    // Build a `CoseSign1` object.
+    let protected = coset::HeaderBuilder::new()
+        .algorithm(iana::Algorithm::ES384)
+        .build();
+    let unprotected = coset::HeaderBuilder::new()
+        .key_id((*b"AsymmetricECDSA384").into())
+        .build();
+    let sign1 = coset::CoseSign1Builder::new()
+        .protected(protected)
+        .unprotected(unprotected)
+        .payload(claims.clone().to_vec()?)
+        .create_signature(aad, |data| stage0_signer.sign(data))
+        .build();
+    sign1.to_vec()
+}
+
+pub fn generate_stage1_attestation() {
```

Comment:
Please add a doc comment for public functions.

Reply:
done